src/backbone_router/nd_proxy.cpp

```diff
@@ -0,0 +1,575 @@
+                 error = OTBR_ERROR_ERRNO);
+#else  // __NetBSD__ || __FreeBSD__ || __APPLE__
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_IP, IP_BOUND_IF, mBackboneIfName.c_str(), mBackboneIfName.size()),
+                 error = OTBR_ERROR_ERRNO);
+#endif // __linux__
+
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_IPV6, IPV6_RECVPKTINFO, &on, sizeof(on)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_IPV6, IPV6_RECVHOPLIMIT, &on, sizeof(on)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_IPV6, IPV6_MULTICAST_HOPS, &hops, sizeof(hops)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_IPV6, IPV6_UNICAST_HOPS, &hops, sizeof(hops)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+
+    ICMP6_FILTER_SETBLOCKALL(&filter);
+    ICMP6_FILTER_SETPASS(ND_NEIGHBOR_SOLICIT, &filter);
+
+    VerifyOrExit(setsockopt(mIcmp6RawSock, IPPROTO_ICMPV6, ICMP6_FILTER, &filter, sizeof(filter)) == 0,
+                 error = OTBR_ERROR_ERRNO);
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        FiniIcmp6RawSocket();
+    }
+
+    return error;
+}
+
+void NdProxyManager::FiniIcmp6RawSocket(void)
+{
+    if (mIcmp6RawSock != -1)
+    {
+        close(mIcmp6RawSock);
+        mIcmp6RawSock = -1;
+    }
+}
+
+otbrError NdProxyManager::InitNetfilterQueue(void)
+{
+    otbrError error = OTBR_ERROR_ERRNO;
+
+    VerifyOrExit((mNfqHandler = nfq_open()) != nullptr);
+    VerifyOrExit(nfq_unbind_pf(mNfqHandler, AF_INET6) >= 0);
+    VerifyOrExit(nfq_bind_pf(mNfqHandler, AF_INET6) >= 0);
+
+    VerifyOrExit((mNfqQueueHandler = nfq_create_queue(mNfqHandler, 88, HandleNetfilterQueue, this)) != nullptr);
+    VerifyOrExit(nfq_set_mode(mNfqQueueHandler, NFQNL_COPY_PACKET, 0xffff) >= 0);
+    VerifyOrExit((mUnicastNsQueueSock = nfq_fd(mNfqHandler)) >= 0);
+
+    error = OTBR_ERROR_NONE;
+
+exit:
+    otbrLogResult(error, "NdProxyManager: %s", __FUNCTION__);
+
+    if (error != OTBR_ERROR_NONE)
+    {
+        FiniNetfilterQueue();
+    }
+
+    return error;
+}
+
+void NdProxyManager::FiniNetfilterQueue(void)
+{
+    if (mUnicastNsQueueSock != -1)
+    {
+        close(mUnicastNsQueueSock);
+        mUnicastNsQueueSock = -1;
+    }
+
+    if (mNfqQueueHandler != nullptr)
+    {
+        nfq_destroy_queue(mNfqQueueHandler);
+        mNfqQueueHandler = nullptr;
+    }
+
+    if (mNfqHandler != nullptr)
+    {
+        nfq_close(mNfqHandler);
+        mNfqHandler = nullptr;
+    }
+}
+
+int NdProxyManager::HandleNetfilterQueue(struct nfq_q_handle *aNfQueueHandler,
+                                         struct nfgenmsg *    aNfMsg,
+                                         struct nfq_data *    aNfData,
+                                         void *               aContext)
+{
+    return reinterpret_cast<NdProxyManager *>(aContext)->HandleNetfilterQueue(aNfQueueHandler, aNfMsg, aNfData);
```

Comment:
```suggestion
    return static_cast<NdProxyManager *>(aContext)->HandleNetfilterQueue(aNfQueueHandler, aNfMsg, aNfData);
```

Reply:
What's the advantage of using `static_cast` compared to `reinterpret_cast`.

Reply:
I think it is generally desirable to use the most strict form that is possible.